bounded_sum_known_n/BoundedSumKnownN.tex

```diff
@@ -0,0 +1,394 @@
+\end{itemize}
+
+These changes will soon be applied to the Rust code. The pseudocode also uses the variable \texttt{n} instead of Rust's \texttt{size}, which is more conventional for proof writing.
+
+
+\section{Algorithm Implementation}
+\subsection{Code in Rust}
+The current OpenDP library contains the transformation \texttt{make\_bounded\_sum\_n} implementing the bounded sum function with known $n$. This is defined in lines 29-49 of the file \texttt{mod.rs} in the Git repository\footnote{Updated on September 20, 2021.} (\url{https://github.com/opendp/opendp/blob/8bbb0fab1da9b86c50235a36d7026189be43b1ab/rust/opendp/src/trans/sum/mod.rs#L29-L49}).
+
+%\begin{figure}[ht]
+    %\includegraphics[width=15cm]{bounded_sum_2.png}
+    %\centering
+    %\label{fig:code}
+%\end{figure}
+
+\begin{figure}[ht]
+    \includegraphics[width=15cm]{bounded_sum_n_new.png}
+    \centering
+    \label{fig:code}
+\end{figure}
+
+\subsection{Pseudocode in Python}
+\label{sec:pseudocode}
+We present a simplified Python-like pseudocode of the Rust implementation below. The necessary definitions for the pseudocode can be found in Section \ref{sec:versioned-docs}.
+
+\subsubsection*{Preconditions}
+To ensure the correctness of the output, we require the following preconditions:
+
+\begin{itemize}
+    \item \textbf{User-specified types:}
+    \begin{itemize}
+        \item Variable \texttt{n} must be of type \texttt{usize}.
+        \item Type \texttt{T} must have traits \texttt{DistanceConstant(IntDistance)}, \texttt{TotalOrd}, \texttt{CheckedMul}, \texttt{CheckNull}, \texttt{Sum(Output=T)}, \texttt{Sub(Output=T)},  and \texttt{ExactIntCast(usize)}.
+        \item \texttt{IntDistance} must have trait \texttt{InfCast(T)}. (Note that this bullet point is not needed in this proof, but it is needed in the code so a hint can be constructed; otherwise a binary search would be needed to construct the hint.)
+        \item Variables \texttt{U} and \texttt{L} must be of type \texttt{T}, and we must have $\texttt{L} \leq \texttt{U}$.
+
+\end{itemize}
+\end{itemize}
+
+\subsubsection*{Postconditions}
+\begin{itemize}
+    \item Either a valid \texttt{Transformation} is returned or an error is returned.
+\end{itemize}
+
+\begin{lstlisting}[language=Python, escapechar=|]
+def MakeBoundedSumN(L: T, U: T, n: usize):
+    input_domain = SizedDomain(VectorDomain(IntervalDomain(L, U)), n)
+    output_domain = AllDomain(T)
+    input_metric = SymmetricDistance()|\label{code:symm-dist}|
+    output_metric = AbsoluteDistance(T)|\label{code:abs-dist}|
+    
+    n_ = exact_int_cast(n, T)|\label{code:exact-int-cast}|
+    if checked_mul(L, n_).is_none or checked_mul(U, n_).is_none: 
+        raise Exception('Potential overflow') |\label{line:max}|
+    
+    def relation(d_in: u32, d_out: T) -> bool: |\label{line:rel}|
+        if checked_mul(2, d_out).is_none or checked_mul(d_in, U-L).is_none
```

Comment:
should d_in be replaced with inf_cast(d_in,T), so that both arguments to checked_mul are of type T?

Reply:
That is a good point. I changed it, but I am leaving this comment "unresolved" in case other people have different ideas.